tree/treeplayer/src/TArrowDS.cxx

```diff
@@ -0,0 +1,324 @@
+   if (!field) {
+      std::string msg = "The dataset does not have column ";
+      msg += colName;
+      throw std::runtime_error(msg);
+   }
+   TDFTypeNameGetter typeGetter;
+   auto status = field->type()->Accept(&typeGetter);
+   if (status.ok() == false) {
+      std::string msg = "TArrowDS does not support a column of type ";
+      msg += field->type()->name();
+      throw std::runtime_error(msg);
+   }
+   return typeGetter.result();
+}
+
+bool TArrowDS::HasColumn(std::string_view colName) const
+{
+   auto field = fTable->schema()->GetFieldByName(std::string(colName));
+   if (!field) {
+      return false;
+   }
+   return true;
+}
+
+void TArrowDS::SetEntry(unsigned int slot, ULong64_t entry)
+{
+   for (auto link : fGetterIndex) {
+      auto column = fTable->column(link.first);
+      auto &getter = fValueGetters[link.second];
+      getter.SetEntry(slot, entry);
+   }
+}
+
+void TArrowDS::InitSlot(unsigned int slot, ULong64_t entry)
+{
+   for (auto link : fGetterIndex) {
+      auto column = fTable->column(link.first);
+      auto &getter = fValueGetters[link.second];
+      getter.UncachedSlotLookup(slot, entry);
+   }
+}
+
+void TArrowDS::SetNSlots(unsigned int nSlots)
+{
+   // We dump all the previous getters structures and we rebuild it.
+   auto nColumns = fGetterIndex.size();
+   fValueGetters.clear();
+   fValueGetters.reserve(nColumns);
+   for (size_t ci = 0; ci != nColumns; ++ci) {
+      auto chunkedArray = fTable->column(fGetterIndex[ci].first)->data();
+      fValueGetters.push_back(ValueGetter{nSlots, chunkedArray->chunks()});
+   }
+
+   // We use the same logic as the ROOTDS.
+   auto splitInEqualRanges = [&nSlots = fNSlots, &ranges = fEntryRanges ](int nRecords, unsigned int newNSlots)
+   {
+      ranges.clear();
+      nSlots = newNSlots;
+      const auto chunkSize = nRecords / nSlots;
+      const auto remainder = 1U == nSlots ? 0 : nRecords % nSlots;
+      auto start = 0UL;
+      auto end = 0UL;
+      for (auto i : ROOT::TSeqU(nSlots)) {
+         start = end;
+         end += chunkSize;
+         ranges.emplace_back(start, end);
+         (void)i;
+      }
+      ranges.back().second += remainder;
+   };
+
+   auto getNRecords = [&table = fTable, &columns = fColumnNames ]()->int
+   {
+      auto index = table->schema()->GetFieldIndex(columns[0]);
+      return table->column(index)->length();
+   };
+
+   auto nRecords = getNRecords();
+   splitInEqualRanges(nRecords, nSlots);
+}
+
+/// This needs to return a pointer to the pointer each value getter
+/// will point to.
+std::vector<void *> TArrowDS::GetColumnReadersImpl(std::string_view colName, const std::type_info &)
+{
+   auto findGetterIndex = [&index = fGetterIndex](unsigned int column)
+   {
+      for (auto &entry : index) {
+         if (entry.first == column) {
+            return entry.second;
+         }
+      }
+      throw std::runtime_error("No column found at index " + std::to_string(column));
+   };
```

Comment:
Because it makes the purpose of the code below clearer:

```C++
const int columnIdx = fTable->schema()->GetFieldIndex(std::string(colName));
const int getterIdx = findGetterIndex(columnIdx);
assert(getterIdx != -1);
assert(getterIdx < fValueGetters.size());
return fValueGetters[getterIdx].slotPtrs();
```